File: src/usage.js
const { mutateDb, now } = require("./store");

function getUsageStats(db, { userId = null, days = 30 } = {}) {
  const since = new Date();
  since.setDate(since.getDate() - days);
  const sinceIso = since.toISOString();

  const inRange = (item) => {
    if (userId && item.userId !== userId) return false;
    return item.timestamp >= sinceIso;
  };
  const usersById = new Map((db.users || []).map((user) => [user.id, user]));
  if (!usersById.has("__admin__")) {
    usersById.set("__admin__", { id: "__admin__", name: "Administrator", username: "admin" });
  }
  const normalizeRecord = (item, legacy = false) => {
    const owner = usersById.get(item.userId) || {};
    const promptTokens = Number(item.promptTokens || 0);
    const completionTokens = Number(item.completionTokens || 0);
    const totalTokens = Number(item.totalTokens || promptTokens + completionTokens || 0);

    return {
      id: item.id,
      userId: item.userId,
      userName: item.userName || owner.name || "",
      username: item.username || owner.username || "",
      apiKeyId: item.apiKeyId || "",
      apiKeyName: item.apiKeyName || "",
      apiKeyPreview: item.apiKeyPreview || "",
      providerId: item.providerId || "",
      providerName: item.providerName || "",
      model: item.model || "unknown",
      upstreamModel: item.upstreamModel || "",
      endpoint: item.endpoint || "",
      method: item.method || "",
      status: Number(item.status || (legacy ? 200 : 0)),
      ok: legacy ? true : Boolean(item.ok),
      stream: Boolean(item.stream),
      durationMs: Number(item.durationMs || 0),
      promptTokens,
      completionTokens,
      totalTokens,
      cachedTokens: Number(item.cachedTokens || 0),
      cacheCreationTokens: Number(item.cacheCreationTokens || 0),
      cacheMissTokens: Number(item.cacheMissTokens || 0),
      reasoningTokens: Number(item.reasoningTokens || 0),
      errorCode: item.errorCode || "",
      errorMessage: item.errorMessage || "",
      timestamp: item.timestamp
    };
  };
  const requestRecords = (db.requestLogs || []).filter(inRange).map((item) => normalizeRecord(item));
  const legacyRecords = (db.tokenUsage || []).filter(inRange).map((item) => normalizeRecord(item, true));
  const records = [...legacyRecords, ...requestRecords].sort((a, b) =>
    String(a.timestamp || "").localeCompare(String(b.timestamp || ""))
  );

  const totalPrompt = records.reduce((sum, item) => sum + (item.promptTokens || 0), 0);
  const totalCompletion = records.reduce((sum, item) => sum + (item.completionTokens || 0), 0);
  const totalTokens = records.reduce((sum, item) => sum + (item.totalTokens || 0), 0);
  const totalCachedTokens = records.reduce((sum, item) => sum + (item.cachedTokens || 0), 0);
  const totalCacheCreationTokens = records.reduce((sum, item) => sum + (item.cacheCreationTokens || 0), 0);
  const totalCacheMissTokens = records.reduce((sum, item) => sum + (item.cacheMissTokens || 0), 0);
  const totalReasoningTokens = records.reduce((sum, item) => sum + (item.reasoningTokens || 0), 0);
  const failedRequests = records.filter((item) => !item.ok).length;

  const byUser = {};
  const byModel = {};
  const byDay = {};
  const byApiKey = {};

  for (const item of records) {
    if (!byUser[item.userId]) {
      byUser[item.userId] = {
        userId: item.userId,
        userName: item.userName,
        username: item.username,
        promptTokens: 0,
        completionTokens: 0,
        totalTokens: 0,
        cachedTokens: 0,
        cacheCreationTokens: 0,
        cacheMissTokens: 0,
        reasoningTokens: 0,
        requests: 0,
        failedRequests: 0
      };
    }
    byUser[item.userId].promptTokens += item.promptTokens || 0;
    byUser[item.userId].completionTokens += item.completionTokens || 0;
    byUser[item.userId].totalTokens += item.totalTokens || 0;
    byUser[item.userId].cachedTokens += item.cachedTokens || 0;
    byUser[item.userId].cacheCreationTokens += item.cacheCreationTokens || 0;
    byUser[item.userId].cacheMissTokens += item.cacheMissTokens || 0;
    byUser[item.userId].reasoningTokens += item.reasoningTokens || 0;
    byUser[item.userId].requests += 1;
    if (!item.ok) byUser[item.userId].failedRequests += 1;

    const apiKeyKey = `${item.userId}:${item.apiKeyId || item.apiKeyPreview || "unknown"}`;
    if (!byApiKey[apiKeyKey]) {
      byApiKey[apiKeyKey] = {
        userId: item.userId,
        userName: item.userName,
        username: item.username,
        apiKeyId: item.apiKeyId,
        apiKeyName: item.apiKeyName || "未知 Key",
        apiKeyPreview: item.apiKeyPreview || "",
        promptTokens: 0,
        completionTokens: 0,
        totalTokens: 0,
        cachedTokens: 0,
        cacheCreationTokens: 0,
        cacheMissTokens: 0,
        reasoningTokens: 0,
        requests: 0,
        failedRequests: 0
      };
    }
    byApiKey[apiKeyKey].promptTokens += item.promptTokens || 0;
    byApiKey[apiKeyKey].completionTokens += item.completionTokens || 0;
    byApiKey[apiKeyKey].totalTokens += item.totalTokens || 0;
    byApiKey[apiKeyKey].cachedTokens += item.cachedTokens || 0;
    byApiKey[apiKeyKey].cacheCreationTokens += item.cacheCreationTokens || 0;
    byApiKey[apiKeyKey].cacheMissTokens += item.cacheMissTokens || 0;
    byApiKey[apiKeyKey].reasoningTokens += item.reasoningTokens || 0;
    byApiKey[apiKeyKey].requests += 1;
    if (!item.ok) byApiKey[apiKeyKey].failedRequests += 1;

    const modelKey = item.model || "unknown";
    if (!byModel[modelKey]) {
      byModel[modelKey] = {
        model: modelKey,
        promptTokens: 0,
        completionTokens: 0,
        totalTokens: 0,
        cachedTokens: 0,
        cacheCreationTokens: 0,
        cacheMissTokens: 0,
        reasoningTokens: 0,
        requests: 0,
        failedRequests: 0
      };
    }
    byModel[modelKey].promptTokens += item.promptTokens || 0;
    byModel[modelKey].completionTokens += item.completionTokens || 0;
    byModel[modelKey].totalTokens += item.totalTokens || 0;
    byModel[modelKey].cachedTokens += item.cachedTokens || 0;
    byModel[modelKey].cacheCreationTokens += item.cacheCreationTokens || 0;
    byModel[modelKey].cacheMissTokens += item.cacheMissTokens || 0;
    byModel[modelKey].reasoningTokens += item.reasoningTokens || 0;
    byModel[modelKey].requests += 1;
    if (!item.ok) byModel[modelKey].failedRequests += 1;

    const day = item.timestamp.slice(0, 10);
    if (!byDay[day]) {
      byDay[day] = {
        day,
        promptTokens: 0,
        completionTokens: 0,
        totalTokens: 0,
        cachedTokens: 0,
        cacheCreationTokens: 0,
        cacheMissTokens: 0,
        reasoningTokens: 0,
        requests: 0,
        failedRequests: 0
      };
    }
    byDay[day].promptTokens += item.promptTokens || 0;
    byDay[day].completionTokens += item.completionTokens || 0;
    byDay[day].totalTokens += item.totalTokens || 0;
    byDay[day].cachedTokens += item.cachedTokens || 0;
    byDay[day].cacheCreationTokens += item.cacheCreationTokens || 0;
    byDay[day].cacheMissTokens += item.cacheMissTokens || 0;
    byDay[day].reasoningTokens += item.reasoningTokens || 0;
    byDay[day].requests += 1;
    if (!item.ok) byDay[day].failedRequests += 1;
  }

  const recentRequests = records.slice(-100).reverse();

  const hourSince = new Date();
  hourSince.setHours(hourSince.getHours() - 24);
  const hourSinceIso = hourSince.toISOString();
  const byHour = {};
  for (const item of records) {
    if (item.timestamp < hourSinceIso) continue;
    const hour = item.timestamp.slice(0, 13) + ":00";
    if (!byHour[hour]) {
      byHour[hour] = {
        hour,
        promptTokens: 0,
        completionTokens: 0,
        totalTokens: 0,
        requests: 0
      };
    }
    byHour[hour].promptTokens += item.promptTokens || 0;
    byHour[hour].completionTokens += item.completionTokens || 0;
    byHour[hour].totalTokens += item.totalTokens || 0;
    byHour[hour].requests += 1;
  }

  return {
    totalPromptTokens: totalPrompt,
    totalCompletionTokens: totalCompletion,
    totalTokens,
    totalCachedTokens,
    totalCacheCreationTokens,
    totalCacheMissTokens,
    totalReasoningTokens,
    requests: records.length,
    failedRequests,
    byUser: Object.values(byUser).sort((a, b) => b.totalTokens - a.totalTokens),
    byApiKey: Object.values(byApiKey).sort((a, b) => b.totalTokens - a.totalTokens),
    byModel: Object.values(byModel).sort((a, b) => b.totalTokens - a.totalTokens),
    byDay: Object.values(byDay).sort((a, b) => a.day.localeCompare(b.day)),
    byHour: Object.values(byHour).sort((a, b) => a.hour.localeCompare(b.hour)),
    recent: recentRequests,
    recentRequests
  };
}

module.exports = {
  getUsageStats
};

File: backend/usage/usage.go
package usage

import (
	"sort"
	"time"

	"sapi/billing"
	"sapi/models"
	"sapi/store"
)

type UsageStats struct {
	TotalPromptTokens        int                 `json:"totalPromptTokens"`
	TotalCompletionTokens    int                 `json:"totalCompletionTokens"`
	TotalTokens              int                 `json:"totalTokens"`
	TotalCachedTokens        int                 `json:"totalCachedTokens"`
	TotalCacheCreationTokens int                 `json:"totalCacheCreationTokens"`
	TotalCacheMissTokens     int                 `json:"totalCacheMissTokens"`
	TotalReasoningTokens     int                 `json:"totalReasoningTokens"`
	TotalCostUSD             float64             `json:"totalCostUsd"`
	TotalCostCNY             float64             `json:"totalCostCny"`
	TotalBillableMicrounits  int64               `json:"totalBillableMicrounits"`
	Requests                 int                 `json:"requests"`
	FailedRequests           int                 `json:"failedRequests"`
	ByUser                   []UserUsageStats    `json:"byUser"`
	ByAPIKey                 []APIKeyUsageStats  `json:"byApiKey"`
	ByModel                  []ModelUsageStats   `json:"byModel"`
	ByDay                    []DayUsageStats     `json:"byDay"`
	ByHour                   []HourUsageStats    `json:"byHour"`
	Recent                   []models.RequestLog `json:"recent"`
	RecentRequests           []models.RequestLog `json:"recentRequests"`
}

type UserUsageStats struct {
	UserID              string  `json:"userId"`
	UserName            string  `json:"userName"`
	Username            string  `json:"username"`
	PromptTokens        int     `json:"promptTokens"`
	CompletionTokens    int     `json:"completionTokens"`
	TotalTokens         int     `json:"totalTokens"`
	CachedTokens        int     `json:"cachedTokens"`
	CacheCreationTokens int     `json:"cacheCreationTokens"`
	CacheMissTokens     int     `json:"cacheMissTokens"`
	ReasoningTokens     int     `json:"reasoningTokens"`
	CostUSD             float64 `json:"costUsd"`
	CostCNY             float64 `json:"costCny"`
	BillableMicrounits  int64   `json:"billableMicrounits"`
	Requests            int     `json:"requests"`
	FailedRequests      int     `json:"failedRequests"`
}

type APIKeyUsageStats struct {
	UserID              string  `json:"userId"`
	UserName            string  `json:"userName"`
	Username            string  `json:"username"`
	APIKeyID            string  `json:"apiKeyId"`
	APIKeyName          string  `json:"apiKeyName"`
	APIKeyPreview       string  `json:"apiKeyPreview"`
	PromptTokens        int     `json:"promptTokens"`
	CompletionTokens    int     `json:"completionTokens"`
	TotalTokens         int     `json:"totalTokens"`
	CachedTokens        int     `json:"cachedTokens"`
	CacheCreationTokens int     `json:"cacheCreationTokens"`
	CacheMissTokens     int     `json:"cacheMissTokens"`
	ReasoningTokens     int     `json:"reasoningTokens"`
	CostUSD             float64 `json:"costUsd"`
	CostCNY             float64 `json:"costCny"`
	BillableMicrounits  int64   `json:"billableMicrounits"`
	Requests            int     `json:"requests"`
	FailedRequests      int     `json:"failedRequests"`
}

type ModelUsageStats struct {
	Model               string  `json:"model"`
	PromptTokens        int     `json:"promptTokens"`
	CompletionTokens    int     `json:"completionTokens"`
	TotalTokens         int     `json:"totalTokens"`
	CachedTokens        int     `json:"cachedTokens"`
	CacheCreationTokens int     `json:"cacheCreationTokens"`
	CacheMissTokens     int     `json:"cacheMissTokens"`
	ReasoningTokens     int     `json:"reasoningTokens"`
	CostUSD             float64 `json:"costUsd"`
	CostCNY             float64 `json:"costCny"`
	BillableMicrounits  int64   `json:"billableMicrounits"`
	Requests            int     `json:"requests"`
	FailedRequests      int     `json:"failedRequests"`
}

type DayUsageStats struct {
	Day                 string  `json:"day"`
	PromptTokens        int     `json:"promptTokens"`
	CompletionTokens    int     `json:"completionTokens"`
	TotalTokens         int     `json:"totalTokens"`
	CachedTokens        int     `json:"cachedTokens"`
	CacheCreationTokens int     `json:"cacheCreationTokens"`
	CacheMissTokens     int     `json:"cacheMissTokens"`
	ReasoningTokens     int     `json:"reasoningTokens"`
	CostUSD             float64 `json:"costUsd"`
	CostCNY             float64 `json:"costCny"`
	BillableMicrounits  int64   `json:"billableMicrounits"`
	Requests            int     `json:"requests"`
	FailedRequests      int     `json:"failedRequests"`
}

type HourUsageStats struct {
	Hour               string  `json:"hour"`
	PromptTokens       int     `json:"promptTokens"`
	CompletionTokens   int     `json:"completionTokens"`
	TotalTokens        int     `json:"totalTokens"`
	CostCNY            float64 `json:"costCny"`
	BillableMicrounits int64   `json:"billableMicrounits"`
	Requests           int     `json:"requests"`
}

func GetUsageStats(db *models.Database, userID string, days int) *UsageStats {
	since := time.Now().AddDate(0, 0, -days)

	usersByID := make(map[string]*models.User)
	for i := range db.Users {
		usersByID[db.Users[i].ID] = &db.Users[i]
	}
	usersByID[models.AdminVirtualUserID] = &models.User{
		ID: models.AdminVirtualUserID, Name: "Administrator", Username: "admin",
	}

	records := store.RequestLogsSince(db, since, userID, 50000)

	sort.Slice(records, func(i, j int) bool {
		return records[i].Timestamp < records[j].Timestamp
	})

	stats := &UsageStats{}
	byUser := make(map[string]*UserUsageStats)
	byAPIKey := make(map[string]*APIKeyUsageStats)
	byModel := make(map[string]*ModelUsageStats)
	byDay := make(map[string]*DayUsageStats)
	byHour := make(map[string]*HourUsageStats)

	for _, item := range records {
		cost := costForItem(db, item)
		stats.Requests++
		stats.TotalPromptTokens += item.PromptTokens
		stats.TotalCompletionTokens += item.CompletionTokens
		stats.TotalTokens += item.TotalTokens
		stats.TotalCachedTokens += item.CachedTokens
		stats.TotalCacheCreationTokens += item.CacheCreationTokens
		stats.TotalCacheMissTokens += item.CacheMissTokens
		stats.TotalReasoningTokens += item.ReasoningTokens
		stats.TotalCostUSD += cost.CostUSD
		stats.TotalCostCNY += cost.CostCNY
		stats.TotalBillableMicrounits += cost.BillableMicrounits
		if !item.OK {
			stats.FailedRequests++
		}

		if _, ok := byUser[item.UserID]; !ok {
			owner := usersByID[item.UserID]
			name := ""
			username := ""
			if owner != nil {
				name = owner.Name
				username = owner.Username
			}
			byUser[item.UserID] = &UserUsageStats{
				UserID: item.UserID, UserName: name, Username: username,
			}
		}
		uu := byUser[item.UserID]
		uu.PromptTokens += item.PromptTokens
		uu.CompletionTokens += item.CompletionTokens
		uu.TotalTokens += item.TotalTokens
		uu.CachedTokens += item.CachedTokens
		uu.CacheCreationTokens += item.CacheCreationTokens
		uu.CacheMissTokens += item.CacheMissTokens
		uu.ReasoningTokens += item.ReasoningTokens
		uu.CostUSD += cost.CostUSD
		uu.CostCNY += cost.CostCNY
		uu.BillableMicrounits += cost.BillableMicrounits
		uu.Requests++
		if !item.OK {
			uu.FailedRequests++
		}

		keyKey := item.UserID + ":" + item.APIKeyID
		if _, ok := byAPIKey[keyKey]; !ok {
			byAPIKey[keyKey] = &APIKeyUsageStats{
				UserID: item.UserID, UserName: item.UserName, Username: item.Username,
				APIKeyID: item.APIKeyID, APIKeyName: item.APIKeyName, APIKeyPreview: item.APIKeyPreview,
			}
		}
		ak := byAPIKey[keyKey]
		ak.PromptTokens += item.PromptTokens
		ak.CompletionTokens += item.CompletionTokens
		ak.TotalTokens += item.TotalTokens
		ak.CachedTokens += item.CachedTokens
		ak.CacheCreationTokens += item.CacheCreationTokens
		ak.CacheMissTokens += item.CacheMissTokens
		ak.ReasoningTokens += item.ReasoningTokens
		ak.CostUSD += cost.CostUSD
		ak.CostCNY += cost.CostCNY
		ak.BillableMicrounits += cost.BillableMicrounits
		ak.Requests++
		if !item.OK {
			ak.FailedRequests++
		}

		modelKey := item.Model
		if modelKey == "" {
			modelKey = "unknown"
		}
		if _, ok := byModel[modelKey]; !ok {
			byModel[modelKey] = &ModelUsageStats{Model: modelKey}
		}
		bm := byModel[modelKey]
		bm.PromptTokens += item.PromptTokens
		bm.CompletionTokens += item.CompletionTokens
		bm.TotalTokens += item.TotalTokens
		bm.CachedTokens += item.CachedTokens
		bm.CacheCreationTokens += item.CacheCreationTokens
		bm.CacheMissTokens += item.CacheMissTokens
		bm.ReasoningTokens += item.ReasoningTokens
		bm.CostUSD += cost.CostUSD
		bm.CostCNY += cost.CostCNY
		bm.BillableMicrounits += cost.BillableMicrounits
		bm.Requests++
		if !item.OK {
			bm.FailedRequests++
		}

		day := ""
		if len(item.Timestamp) >= 10 {
			day = item.Timestamp[:10]
		}
		if _, ok := byDay[day]; !ok {
			byDay[day] = &DayUsageStats{Day: day}
		}
		bd := byDay[day]
		bd.PromptTokens += item.PromptTokens
		bd.CompletionTokens += item.CompletionTokens
		bd.TotalTokens += item.TotalTokens
		bd.CachedTokens += item.CachedTokens
		bd.CacheCreationTokens += item.CacheCreationTokens
		bd.CacheMissTokens += item.CacheMissTokens
		bd.ReasoningTokens += item.ReasoningTokens
		bd.CostUSD += cost.CostUSD
		bd.CostCNY += cost.CostCNY
		bd.BillableMicrounits += cost.BillableMicrounits
		bd.Requests++
		if !item.OK {
			bd.FailedRequests++
		}

		hourSince := time.Now().Add(-24 * time.Hour).UTC().Format(time.RFC3339)
		if item.Timestamp >= hourSince && len(item.Timestamp) >= 13 {
			hour := item.Timestamp[:13] + ":00"
			if _, ok := byHour[hour]; !ok {
				byHour[hour] = &HourUsageStats{Hour: hour}
			}
			bh := byHour[hour]
			bh.PromptTokens += item.PromptTokens
			bh.CompletionTokens += item.CompletionTokens
			bh.TotalTokens += item.TotalTokens
			bh.CostCNY += cost.CostCNY
			bh.BillableMicrounits += cost.BillableMicrounits
			bh.Requests++
		}
	}

	for _, v := range byUser {
		stats.ByUser = append(stats.ByUser, *v)
	}
	sort.Slice(stats.ByUser, func(i, j int) bool {
		return stats.ByUser[i].TotalTokens > stats.ByUser[j].TotalTokens
	})

	for _, v := range byAPIKey {
		stats.ByAPIKey = append(stats.ByAPIKey, *v)
	}
	sort.Slice(stats.ByAPIKey, func(i, j int) bool {
		return stats.ByAPIKey[i].TotalTokens > stats.ByAPIKey[j].TotalTokens
	})

	for _, v := range byModel {
		stats.ByModel = append(stats.ByModel, *v)
	}
	sort.Slice(stats.ByModel, func(i, j int) bool {
		return stats.ByModel[i].TotalTokens > stats.ByModel[j].TotalTokens
	})

	for _, v := range byDay {
		stats.ByDay = append(stats.ByDay, *v)
	}
	sort.Slice(stats.ByDay, func(i, j int) bool {
		return stats.ByDay[i].Day < stats.ByDay[j].Day
	})

	for _, v := range byHour {
		stats.ByHour = append(stats.ByHour, *v)
	}
	sort.Slice(stats.ByHour, func(i, j int) bool {
		return stats.ByHour[i].Hour < stats.ByHour[j].Hour
	})

	recentStart := max(0, len(records)-100)
	recent := records[recentStart:]
	for i := len(recent)/2 - 1; i >= 0; i-- {
		opp := len(recent) - 1 - i
		recent[i], recent[opp] = recent[opp], recent[i]
	}
	stats.Recent = recent
	stats.RecentRequests = recent

	return stats
}

func costForItem(db *models.Database, item models.RequestLog) billing.CostBreakdown {
	if item.BillableMicrounits > 0 || item.CostUSD > 0 || item.CostCNY > 0 {
		return billing.CostBreakdown{
			CostUSD:            item.CostUSD,
			CostCNY:            item.CostCNY,
			BillableMicrounits: item.BillableMicrounits,
		}
	}
	return billing.CalculateRequestCost(db, item)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

type numeric interface{ int | float64 }

func maxOf[T numeric](a, b T) T {
	if a > b {
		return a
	}
	return b
}
